src/request.rs

```diff
@@ -128,20 +130,34 @@ impl<State> Context<State> {
         Ok(serde_json::from_slice(&body_bytes).map_err(|_| std::io::ErrorKind::InvalidData)?)
     }
 
-    /// Remove ownership of the request body, replacing it with an empty body.
-    ///
-    /// Used primarily for working directly with the body stream.
-    pub fn take_body(&mut self) -> Body {
-        std::mem::replace(self.request.body_mut(), Body::empty())
+    /// Get the URL querystring.
+    pub fn query<'de, T: Deserialize<'de>>(&'de self) -> Result<T, crate::Error> {
+        let query = self.uri().query();
+        if query.is_none() {
+            return Err(crate::Error::from(http::StatusCode::BAD_REQUEST));
+        }
+        Ok(serde_qs::from_str(query.unwrap())
+            .map_err(|_| crate::Error::from(http::StatusCode::BAD_REQUEST))?)
     }
 
-    /// Access the extensions to the context.
-    pub fn extensions(&self) -> &http::Extensions {
-        self.request.extensions()
+    /// Parse the request body as a form.
+    pub async fn body_form<T: serde::de::DeserializeOwned>(&mut self) -> io::Result<T> {
+        let body = self.body_bytes()
+            .await
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
+        let res = serde_qs::from_bytes(&body)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("could not decode form: {}", e)))?;
+        Ok(res)
     }
+}
 
-    /// Mutably access the extensions to the context.
-    pub fn extensions_mut(&mut self) -> &mut http::Extensions {
-        self.request.extensions_mut()
+impl<State> Read for Request<State> {
```

Comment:
Well, I think libraries should be runtime-independent and we should impl `futures_io::AsyncRead` instead. I know that `async-std` doesn't have a blanket impl now, but looking at the code, having one seems *technically* possible if we ignore breaking changes.

Reply:
nvm, it's just a re-export from `futures-io` -- I personally prefer pulling `futures-io` though, but `async-std` with only `std` feature is okay for me too!